motion_tracker/utils/image_cropping.py

```diff
@@ -0,0 +1,254 @@
+    box_for_crop_x1 = box_for_crop_x_center + 0.5 * box_for_crop_width
+    box_for_crop_y1 = box_for_crop_y_center + 0.5 * box_for_crop_height
+    box_for_crop = Coords(box_for_crop_x0, box_for_crop_y0, box_for_crop_x1, box_for_crop_y1)
+    return box_for_crop
+
+def get_crop_coords(box_coords, img_coords, random_crop):
+    '''Calculate the boundaries for a new crop and the coords of the correct bounding box
+    in that new crop.
+
+    Args:
+    -----
+    box_coords: Coords object with location of bounding box in original img
+    img_coords: Coords (basically the size) of original image
+    random_crop: If True, create stochastic box to determine crop coords.
+                 If False, create crop based on true bounding box
+                 Cropped area is twice as large (and centered) on box either way
+
+    Outputs:
+    --------
+    cropped_area_coords: Coords object with boundaries of the random crop
+    new_box_coords: Coords object with location of correct bounding box in
+        newly cropped image
+    '''
+
+    if random_crop:
+        box_for_crop = get_random_box_to_determine_crop(box_coords)
+    else:
+        box_for_crop = box_coords
+
+    # calc boundary coords of cropped area. Crop_img is twice as big as box_for_crop
+    cropped_area_x0 = max(box_for_crop.x_center - box_for_crop.width, 0)
+    cropped_area_y0 = max(box_for_crop.y_center - box_for_crop.height, 0)
+    cropped_area_x1 = min(box_for_crop.x_center + box_for_crop.width, img_coords.width)
+    cropped_area_y1 = min(box_for_crop.y_center + box_for_crop.height, img_coords.height)
+    cropped_area_coords = Coords(cropped_area_x0, cropped_area_y0,
+                                 cropped_area_x1, cropped_area_y1)
+
+    box_after_crop = update_box_after_crop(box_coords, cropped_area_coords)
+    return cropped_area_coords, box_after_crop
+
+def read_bbox_data(raw_image_dir, parsed_bb_path='work/imagenet/parsed_bb.csv'):
+    '''Return df with data about images and bounding boxes.
+
+    Data captured corresponds to xml files obtained from ImageNet.
+    '''
+
+    # paper says they use .66, but they use search area that is twice bounding box
+    # So, 0.5 may be more appropriate.
+    max_box_frac_of_width =  0.66
+    max_box_frac_of_height = 0.66
+    images_successfully_downloaded = set(os.listdir('data/imagenet/images'))
+    bbox_df = (pd.read_csv('work/imagenet/parsed_bb.csv')
+                    .assign(box_height = lambda df: df.ymax - df.ymin,
+                            box_width = lambda df: df.xmax - df.xmin)
+                    .assign(box_frac_of_height = lambda df: df.box_height / df.height,
+                            box_frac_of_width = lambda df:  df.box_width / df.width)
+                    .query('filename in @images_successfully_downloaded')
+                    .query('box_frac_of_height < @max_box_frac_of_height')
+                    .query('box_frac_of_width < @max_box_frac_of_width'))
+    return bbox_df
+
```

Comment:
Your thinking we would do this, save the output, and then load that saved
file when we start the generator?

That seems like a good idea.

On Fri, Apr 29, 2016 at 6:26 AM, Sean Sall <notifications@github.com> wrote:

> In motion_tracker/utils/image_cropping.py
> <https://github.com/dansbecker/motion-tracking/pull/11#discussion_r61568821>
> :
>
> > +
> > +    # paper says they use .66, but they use search area that is twice bounding box
> > +    # So, 0.5 may be more appropriate.
> > +    max_box_frac_of_width =  0.66
> > +    max_box_frac_of_height = 0.66
> > +    images_successfully_downloaded = set(os.listdir('data/imagenet/images'))
> > +    bbox_df = (pd.read_csv('work/imagenet/parsed_bb.csv')
> > +                    .assign(box_height = lambda df: df.ymax - df.ymin,
> > +                            box_width = lambda df: df.xmax - df.xmin)
> > +                    .assign(box_frac_of_height = lambda df: df.box_height / df.height,
> > +                            box_frac_of_width = lambda df:  df.box_width / df.width)
> > +                    .query('filename in @images_successfully_downloaded')
> > +                    .query('box_frac_of_height < @max_box_frac_of_height')
> > +                    .query('box_frac_of_width < @max_box_frac_of_width'))
> > +    return bbox_df
> > +
>
> What if we did this outside of the generator? At some point when we have
> all the images downloaded that we need to, this becomes a 1 time operation,
> yeah? This will shave a little bit of time off the generator setup.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/dansbecker/motion-tracking/pull/11/files/b6491b5f1cf66cac8d3f02969a7bbd975c96aa59#r61568821>
>